Thread-safe dispatcher table

ilia-cher opened the pull request:
Summary: Allow concurrent read and writes in dispatcher table

Differential Revision: D8722560

ezyang commented on caffe2/core/dispatch/DispatchTable.h:
```diff
@@ -1,104 +1,118 @@
-      // TODO Locking
-      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+  template <class Key_>
+  void emplace(Key_&& key, void* value) {
```
Shouldn't we encourage bulk inserts? Otherwise, O(n) emplaces will cost O(n^2).

ilia-cher replied:
good suggestion, but I think this is outside of the scope for this diff

smessmer replied:
How does it cost `O(n^2)`? The hashmap will only reallocate when crossing a `2^n` size boundary, so inserting should be amortized `O(n)`.

smessmer replied:
The current implementation is overly expensive. We should avoid allocating a new hash map every time, then we'd be amortized `O(n)`. It should be possible to just keep two hash maps around, which are both always identical. One is read from, writing writes to the other one, swaps them, then writes the same change to the first one.

ezyang replied:
Well, it didn't use to be O(n^2), because we weren't copying the hashmap before?

ilia-cher commented:
@pytorchbot retest this please

ilia-cher commented:
@pytorchbot retest this please

Commit:
Thread-safe dispatcher table (#9126)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/9126

Closes https://github.com/pytorch/pytorch/pull/9126

Allow concurrent read and writes in dispatcher table

Differential Revision: D8722560

fbshipit-source-id: 52fe48efa715c3c3f9cd44d23135a30263326605

ilia-cher commented:
@pytorchbot  retest this please
